### PR TITLE
Generalize notification infrastructure, including error notification

### PR DESCRIPTION
**Pull request by bobvanmanen**
## Description

* Generalizes the event notification handling in block I/O, including failure notification.
* Adds AEM download support to the testbed application
* Fixes retry handling in AEM download due to stream errors
* Reject unsupported files for upload (special characters, empty files)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by tmathern — APPROVED** [review]
I start to worry about how complex this is getting (eg. classes for function handling, async loops, etc). 
I also wonder if all the eventing/notifications could be simplified by using state machines, or reactive extensions like [RxJs](https://rxjs.dev/).

File: lib/aem/error-codes.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019 Adobe. All rights reserved.
```
[the remark this answers is not in the document]
Comment (bobvanmanen) [reply lib/aem/error-codes.js:2]: This is existing code from `aem-upload` to ensure compatibility.

File: lib/aem/upload-error.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    /**
+     * Retrieves the upload error's status as an HTTP status code.
+     *
+     * @returns {number} An HTTP status code.
+     */
+    getHttpStatusCode() {
+        const code = this.getCode();
+
+        if (code === errorCodes.ALREADY_EXISTS) {
```
[the remark this answers is not in the document]
Comment (bobvanmanen) [reply lib/aem/upload-error.js:130]: This is existing code from `aem-upload` to ensure compatibility.

**Comment by bobvanmanen** [comment]
> I start to worry about how complex this is getting (eg. classes for function handling, async loops, etc).
> I also wonder if all the eventing/notifications could be simplified by using state machines, or reactive extensions like [RxJs](https://rxjs.dev/).

It would be interesting to see if RxJs could be of benefit, however I've used Rx extensions in the past and it's anything but simple. It brings in a lot of complexity of its own.

The only state kept is whether a given asset failed to transfer, given that it's being concurrently transferred and we don't want to fail the other assets in the queue state is kept so those failed assets are filtered out. 

Other than that, it's just notifications from functions to know where in the pipeline the asset or part of the asset resides. Without, it there would be no visibility in that.

There is an event mapping in `aemupload` and `aemdownload` but that's for backward compatibility with the `aem-upload` api. If there was no backward compatibility requirement, we wouldn't need that.

**Comment by tmathern** [comment]
@bobvanmanen It's just I had the feeling there are a bunch of notifications going around :). 
But it's still manageable for now.

**Pull request merged**